### PR TITLE
tracing: skip reconfiguration if tracer already configured

### PR DESCRIPTION
**Pull request by redloaf**
## 💸 TL;DR
In baseplate.go v2 we configure a global opentracing tracer. Currently, calling `baseplate.New` will override the new tracer. This change prevents the override using a flag in `internalv2compat`.

## 📜 Details
If `internalv2compat.SetV2TracingEnabled(true)` is called before `InitGlobalTracer`, the opentracing global tracer will not be configured, and a message will be logged.

## ✅ Checks

- [x] CI tests (if present) are passing
- [x] Adheres to code style for repo
- [ ] Contributor License Agreement (CLA) completed if not a Reddit employee

### CODE REVIEW

File: internalv2compat/compat.go [hunk 1]
```diff
@@ -63,3 +64,19 @@ type IsHTTP interface {
 type IsThrift interface {
 	isThrift()
 }
+
+var (
+	v2TracingEnabledMutex sync.RWMutex
+	v2TracingEnabled      bool
```
Comment (fishy) [inline internalv2compat/compat.go:70]: use [`atomic.Bool`](https://pkg.go.dev/sync/atomic#Bool) instead?

File: tracing/tracer.go [hunk 2]
```diff
@@ -82,6 +83,14 @@ type Tracer struct {
 // If it fails to do so, UndefinedIP will be used instead,
 // and the error will be logged if logger is non-nil.
 func InitGlobalTracer(cfg Config) error {
+	logger := cfg.Logger
+	if logger == nil {
+		logger = log.NopWrapper
+	}
+	if internalv2compat.V2TracingEnabled() {
+		logger(context.Background(), `v2 tracing is enabled; skipping v0 tracer configuration`)
```
Comment (fishy) [inline tracing/tracer.go:91]: nit:
```suggestion
	if internalv2compat.V2TracingEnabled() {
		cfg.Logger.Log(context.Background(), `v2 tracing is enabled; skipping v0 tracer configuration`)
```
see https://pkg.go.dev/github.com/reddit/baseplate.go/log#Wrapper.Log

File: internalv2compat/compat.go [hunk 3]
```diff
@@ -63,3 +64,19 @@ type IsHTTP interface {
...
+func SetV2TracingEnabled(enabled bool) {
+	v2TracingEnabledMutex.Lock()
+	defer v2TracingEnabledMutex.Unlock()
+	v2TracingEnabled = enabled
+}
+func V2TracingEnabled() bool {
+	v2TracingEnabledMutex.RLock()
+	defer v2TracingEnabledMutex.RUnlock()
+	return v2TracingEnabled
+}
```
Comment (kylelemons) [inline internalv2compat/compat.go:82]: * I don't think you need an RWMutex here -- they are a bit slower, and are only really useful if you _need_ the multiple readers for throughput
* I often like combining the "locked thing" in with the mutex.

Together it could look like

```suggestion
var v2Tracing struct{
  sync.Mutex
	enabled bool
}

func SetV2TracingEnabled(enabled bool) {
	v2Tracing.Lock()
	defer v2Tracing.Unlock()
	v2Tracing.Enabled = enabled
}
func V2TracingEnabled() bool {
	v2Tracing.Lock()
	defer v2Tracing.Unlock()
	return v2Tracing.Enabled
}
```

**Review by kylelemons — APPROVED** [review]

File: tracing/tracer.go [hunk 4]
```diff
@@ -12,6 +12,7 @@ import (
 	"github.com/opentracing/opentracing-go"
 
 	"github.com/reddit/baseplate.go/detach"
+	"github.com/reddit/baseplate.go/internalv2compat"
```
Comment (kylelemons) [inline tracing/tracer.go:15]: need a lint ignore for SA1019

File: tracing/tracer.go [hunk 5]
```diff
@@ -102,13 +111,7 @@ func InitGlobalTracer(cfg Config) error {
 
 	tracer.sampleRate = cfg.SampleRate
 	tracer.useHex = cfg.UseHex
-
-	logger := cfg.Logger
-	if logger == nil {
-		logger = log.NopWrapper
-	}
```
Comment (fishy) [inline tracing/tracer.go:109]: and then revert this change.

here we decided intentionally to default to `NopWrapper` instead of `DefaultWrapper`

**Review by fishy — CHANGES_REQUESTED** [review]

File: internalv2compat/compat.go (hunk 3 above)
Comment (redloaf) [reply internalv2compat/compat.go]: Refactored.

File: internalv2compat/compat.go (hunk 1 above)
Comment (redloaf) [reply internalv2compat/compat.go]: I considered it but picked a mutex because this should only be referenced during setup/init, and I feel like the benefit of a memory barrier outweighs the performance benefits of atomics.

**Review by fishy — APPROVED** [review]
